adc/read_vsys/power_status.c

```diff
@@ -0,0 +1,48 @@
+/**
+ * Copyright (c) 2023 Raspberry Pi (Trading) Ltd.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include "stdbool.h"
+#include "hardware/adc.h"
+#include "power_status.h"
+
+#if CYW43_USES_VSYS_GPIO_PIN
+#include "pico/cyw43_arch.h"
+#endif
+
+// Get power status
+// Must have called adc_init
+// On Pico W must also have called cyw43_arch_init
+void power_status(bool *mains_powered, float *voltage_result) {
+#if defined CYW43_WL_GPIO_VBUS_PIN
+    *mains_powered = cyw43_arch_gpio_get(CYW43_WL_GPIO_VBUS_PIN);
+#elif defined PICO_VBUS_GPIO_PIN
+    gpio_set_function(PICO_VBUS_GPIO_PIN, GPIO_FUNC_SIO);
```

Comment:
If this function is changing the GPIO_FUNC, should it also restore it before exiting?

Reply:
I think anything using this pin should change the function before using it. The driver now does this.